Add a workspace drag surface that blocks and bubble get moved to duri…

picklesrus opened the pull request:
…ng a workspace drag.

The surface is translated using translate3d instead of svg's translate attribute so that
the browser does not have to repaint the entire workspace on every mouse move.
This is very similar to the block drag surface.

rachel-fenichel commented on core/css.js:
```diff
@@ -159,6 +159,13 @@ Blockly.Css.CONTENT = [
+    'overflow: visible;',
+    'top: 0;',
+    'left: 0;',
+  '}',
```
Nit: Newline

rachel-fenichel commented on core/touch.js:
```diff
@@ -107,6 +107,9 @@ Blockly.onMouseUp_ = function(e) {
     return;
   }
   Blockly.Touch.clearTouchIdentifier();
+  // Check whether we should do this regardless of drag mode.  See comment on
```
Make this // TODO: (issue number).  Right now it looks like it's saying that the code is doing this check.

rachel-fenichel commented on core/workspace_drag_surface_svg.js:
```diff
@@ -0,0 +1,181 @@
+
+/**
+ * @fileoverview An SVG that floats on top of the workspace.
+ * Blocks are moved into this SVG during a drag, improving performance.
```
Update this comment to say that this lets the blocks be moved with CSS translations instead of SVG translations, which improves performance.

rachel-fenichel commented on core/workspace_drag_surface_svg.js:
```diff
@@ -0,0 +1,181 @@
+
+goog.require('goog.asserts');
+goog.require('goog.math.Coordinate');
+
```
Two newlines after the last goog.require

rachel-fenichel commented on core/workspace_drag_surface_svg.js:
```diff
@@ -0,0 +1,181 @@
+goog.require('goog.math.Coordinate');
+
+/**
+ * Class for a Drag Surface SVG.
```
I think it's okay to have this be the same as the @fileoverview text.  As it is, there's not much information here.

rachel-fenichel commented on core/workspace_drag_surface_svg.js:
```diff
@@ -0,0 +1,181 @@
+
+/**
+ * Class for a Drag Surface SVG.
+ * @param {Element} container Containing element.
```
!Element

rachel-fenichel commented on core/inject.js:
```diff
@@ -55,9 +56,16 @@ Blockly.inject = function(container, opt_options) {
   blockDragSurface.createDom();
-  var workspace = Blockly.createMainWorkspace_(svg, options, blockDragSurface);
+  var workspaceDragSurface = new Blockly.workspaceDragSurfaceSvg(subContainer);
+  workspaceDragSurface.createDom();
```
Would you ever call the workspace drag surface constructor and not immediately call createDom?  If not, the createDom call should be in the constructor.

rachel-fenichel commented on core/workspace_drag_surface_svg.js:
```diff
@@ -0,0 +1,181 @@
+Blockly.workspaceDragSurfaceSvg.prototype.SVG_ = null;
+
+/**
+ * SVG group inside the drag surface. This is where blocks are moved to.
```
"This is where blocks are moved to" doesn't say why, or when, or by whom, or when they move out of this group.  We can talk through wording if you want, but right now this doesn't really provide any information.

rachel-fenichel commented on core/workspace_drag_surface_svg.js:
```diff
@@ -0,0 +1,181 @@
+ */
+Blockly.workspaceDragSurfaceSvg.prototype.container_ = null;
+
+
```
Nit: remove extra newline

rachel-fenichel commented on core/workspace_drag_surface_svg.js:
```diff
@@ -0,0 +1,181 @@
+  }
+
+  /**
+  * FILL dom structure in!
```
Nit: Fill DOM structure in!

rachel-fenichel commented on core/workspace_drag_surface_svg.js:
```diff
@@ -0,0 +1,181 @@
+  this.container_.appendChild(this.SVG_);
+};
+
+
```
Nit: remove newline

rachel-fenichel commented on core/workspace_drag_surface_svg.js:
```diff
@@ -0,0 +1,181 @@
+ * We translate the drag surface instead of the blocks inside the surface
+ * so that the browser avoids repainting the SVG.
+ * Because of this, the drag coordinates must be adjusted by scale.
+ * @param {Number} x X translation for the entire surface
```
{number}, here and on the next line.

rachel-fenichel commented on core/workspace_drag_surface_svg.js:
```diff
@@ -0,0 +1,181 @@
+  // This is a work-around to prevent a bug in Safari, where numbers close to 0
+  // are sometimes reported as something like "2.9842794901924208e-12".
+  // That is incompatible with translate3d, causing bugs.
+  x = x.toFixed(0);
```
Comment doesn't match code: The comment says two decimal points, but the code says no decimal points.

rachel-fenichel commented on core/workspace_drag_surface_svg.js:
```diff
@@ -0,0 +1,181 @@
+  x = x.toFixed(0);
+  y = y.toFixed(0);
+
+  // Ignorning browsers that don't support translate3d at the moment.
```
Nit: Ignoring (spelling)

How are we ignoring those browsers?  I thought that would be where we create the workspace drag surface, but I don't see it in the code.

rachel-fenichel commented on core/workspace_drag_surface_svg.js:
```diff
@@ -0,0 +1,181 @@
+/**
+ * Reports the surface translation in scaled workspace coordinates.
+ * Use this when finishing a drag to return blocks to the correct position.
+ * @return {goog.math.Coordinate} Current translation of the surface
```
@return {!goog.math.Coordinate}

rachel-fenichel commented on core/workspace_drag_surface_svg.js:
```diff
@@ -0,0 +1,181 @@
+};
+
+/**
+ * Clear the  and hide the contents of the  surface;  Move everything back to
```
Clear the what?  Also nit: two spaces between "the" and "surface"

rachel-fenichel commented on core/workspace_drag_surface_svg.js:
```diff
@@ -0,0 +1,181 @@
+ * Clear the  and hide the contents of the  surface;  Move everything back to
+ * newSurface.
+ * @param {!SVGElement} newSurface The element to put the drag surface contents
+ * into.
```
Indent another four spaces.

rachel-fenichel commented on core/workspace_drag_surface_svg.js:
```diff
@@ -0,0 +1,181 @@
+
+  // Reattach the bubble canvas after the blockCanvas.
+  Blockly.utils.insertAfter_(bubbleCanvas, blockCanvas);
+  this.SVG_.style.display = 'none';
```
Newline, then 
// Hide the drag surface
this.SVG_.style.display = 'none';
...

rachel-fenichel commented on core/workspace_drag_surface_svg.js:
```diff
@@ -0,0 +1,181 @@
+};
+
+/**
+ * Set the SVG to have everything on it and show the surface.
```
"everything"?

rachel-fenichel commented on core/workspace_drag_surface_svg.js:
```diff
@@ -0,0 +1,181 @@
+ * @param {!Element} blockCanvas The block canvas <g> element from the workspace.
+ * @param {!Element} bubbleCanvas The <g> element that contains the bubbles.
+ * @param {?Element} previousSibling The element to insert the block canvas &
+   bubble canvas after when it goes back in the dom at the end of a drag.
```
Indent four spaces.

rachel-fenichel commented on core/workspace_svg.js:
```diff
@@ -66,6 +69,13 @@ Blockly.WorkspaceSvg = function(options, opt_blockDragSurface) {
+    this.workspaceDragSurface_ = opt_wsDragSurface;
+  }
+
+  this.useWorkspaceDragSurface_ =
```
We could do the is3dSupported check in inject.js, then just check whether a drag surface has been passed in here.

rachel-fenichel commented on core/workspace_svg.js:
```diff
@@ -581,15 +606,64 @@ Blockly.WorkspaceSvg.prototype.getParentSvg = function() {
   }
 };
 
+
```
Nit: remove extra newline.

picklesrus commented:
Addressed all the comments except for the one about not creating the drag surfaces if supports3d is false. Tomorrow I'll write a hidden iframe test case which may change my mind about whether that's a good idea :)

rachel-fenichel commented on core/inject.js:
```diff
@@ -55,9 +56,15 @@ Blockly.inject = function(container, opt_options) {
+  // Create surfaces for dragging things. These are optimizations
+  // so that the broowser does not repaint during the drag.
   var blockDragSurface = new Blockly.BlockDragSurfaceSvg(subContainer);
   blockDragSurface.createDom();
```
For followup: move createDom into BlockDragSurfaceSvg constructor, as you're doing for the workspace drag surface.

rachel-fenichel commented:
lgtm!  

Expected followups:
-- hidden iframe test case
-- (if test goes well) don't create drag surfaces if supports3d is false
-- move blockDragSurfaceSvg createDom into constructor

picklesrus commented:
Added a hasClass_ utility back and tested that IE11 is working.

rachel-fenichel commented on core/utils.js:
```diff
@@ -82,6 +82,19 @@ Blockly.utils.removeClass = function(element, className) {
+ * @return {boolean} True if class exists, false otherwise.   
+ * @private   
+ */   
+ Blockly.utils.hasClass_ = function(element, className) {    
```
This is not private.

rachel-fenichel commented:
lgtm
